### PR TITLE
19937 add trabslate table

### PR DESCRIPTION
**Pull request by boyjoy1127**
ref #19937.

Changes proposed in this pull request:
- Add TranslateTable to support projection and filter push down to the table rel-node.
- Add optimization rules to support projection and filter push down to the table rel-node.

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/EmptyRowEnumerator.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.shardingsphere.infra.federation.executor.advanced.table;
+
+import org.apache.calcite.linq4j.Enumerator;
+
+/**
+ * Empty row enumerator.
+ */
+public final class EmptyRowEnumerator implements Enumerator<Object[]> {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/EmptyRowEnumerator.java:25]: Can we use the exist EmptyRowEnumerator?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/TranslatableRowEnumerator.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collection;
+
+/**
+ * Translatable row enumerator.
+ */
+@RequiredArgsConstructor
+public final class TranslatableRowEnumerator implements Enumerator<Object[]> {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/TranslatableRowEnumerator.java:34]: Can we use the exist FilterableRowEnumerator?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/FederationTranslatableTable.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                                        final SchemaPlus schema, final String tableName) {
+        throw new UnsupportedOperationException();
+    }
+    
+    @Override
+    public RelNode toRel(
+                         final RelOptTable.ToRelContext context,
+                         final RelOptTable relOptTable) {
+        // Request all fields.
+        final int fieldCount = relOptTable.getRowType().getFieldCount();
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/FederationTranslatableTable.java:108]: Please remove final here.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Relational expression representing a scan.
+ * Like any table scan, it serves as a leaf node of a query tree.
+ */
+@Getter
+public class TranslatableTableScan extends TableScan implements EnumerableRel {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java:57]: Can we rename TranslatableTableScan to ShardingSphereTableScan?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/TranslatableTableScanExecutor.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        if (null != scanContext.getFilterValues()) {
+            builder.filter(createFilters(scanContext.getFilterValues(), builder, table.getColumnNames()));
+        }
+        if (null != scanContext.getProjects()) {
+            builder.project(createProjections(scanContext.getProjects(), builder, table.getColumnNames()));
+        }
+        return builder.build();
+    }
+    
+    private Collection<RexNode> createFilters(final String[] filterValues, final RelBuilder builder, final List<String> columnNames) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/table/TranslatableTableScanExecutor.java:215]: Can we move this logic to the entrance, and remove TranslatableTableScanExecutor?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableFilterRule.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        super(
+                operand(LogicalFilter.class,
+                        operand(TranslatableTableScan.class, none())),
+                relBuilderFactory,
+                "TranslatableFilterRule");
+    }
+    
+    @Override
+    public void onMatch(final RelOptRuleCall call) {
+        final LogicalFilter filter = call.rel(0);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableFilterRule.java:53]: Please remove final modifier.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectFilterRule.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                operand(LogicalProject.class,
+                        operand(LogicalFilter.class,
+                                operand(TranslatableTableScan.class, none()))),
+                relBuilderFactory,
+                "TranslatableProjectFilterRule");
+    }
+    
+    @Override
+    public void onMatch(final RelOptRuleCall call) {
+        final LogicalProject project = call.rel(0);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectFilterRule.java:56]: Please remove final modifier.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectFilterRule.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        final int[] fields = new int[exps.size()];
+        for (int i = 0; i < exps.size(); i++) {
+            final RexNode exp = exps.get(i);
+            if (exp instanceof RexInputRef) {
+                fields[i] = ((RexInputRef) exp).getIndex();
+            } else {
+                return null;
+            }
+        }
+        return fields;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectFilterRule.java:85]: Please rename fields to result.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectRule.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+     * @param relBuilderFactory Builder for relational expressions
+     */
+    public TranslatableProjectRule(final RelBuilderFactory relBuilderFactory) {
+        super(operand(LogicalProject.class, operand(TranslatableTableScan.class, none())),
+                relBuilderFactory, "TranslatableProjectRule");
+    }
+    
+    @Override
+    public void onMatch(final RelOptRuleCall call) {
+        final LogicalProject project = call.rel(0);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectRule.java:49]: Please remove final.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectRule.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        final int[] fields = new int[exps.size()];
+        for (int i = 0; i < exps.size(); i++) {
+            final RexNode exp = exps.get(i);
+            if (exp instanceof RexInputRef) {
+                fields[i] = ((RexInputRef) exp).getIndex();
+            } else {
+                return null;
+            }
+        }
+        return fields;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableProjectRule.java:74]: Please rename fields to result.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    public void register(final RelOptPlanner planner) {
+        planner.addRule(TranslatableProjectFilterRule.INSTANCE);
+        planner.addRule(TranslatableFilterRule.INSTANCE);
+        planner.addRule(TranslatableProjectRule.INSTANCE);
+    }
+    
+    @Override
+    public RelOptCost computeSelfCost(final RelOptPlanner planner,
+                                      final RelMetadataQuery mq) {
+        // Multiply the cost by a factor that makes a scan more attractive if it
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java:140]: Please remove useless java doc.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/FederationExecutorFactory.java [hunk 12]
```diff
@@ -34,20 +34,20 @@ public final class FederationExecutorFactory {
...
+     * @param props              configuration properties
+     * @param jdbcExecutor       jdbc executor
+     * @param eventBusContext    event bus context
      * @return created instance
      */
     public static FederationExecutor newInstance(final String databaseName, final String schemaName, final OptimizerContext optimizerContext,
                                                  final ShardingSphereRuleMetaData globalRuleMetaData, final ConfigurationProperties props, final JDBCExecutor jdbcExecutor,
                                                  final EventBusContext eventBusContext) {
-        // TODO Consider about AdvancedFederationExecutor
+        // TODO Consider about AdvancedFederationExecutor and TranslatableFederationExecutor
```
[the remark this answers is not in the document]
Comment (boyjoy1127) [reply shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/FederationExecutorFactory.java:50]: No, we can't. They use different type of schema and table.

[the remark this answers is not in the document]
Comment (boyjoy1127) [reply shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/FederationExecutorFactory.java:50]: No, we can't. They use different type of schema and table.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/TranslatableFederationExecutor.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Advanced federation executor.
+ */
+public final class TranslatableFederationExecutor implements FederationExecutor {
```
[the remark this answers is not in the document]
Comment (boyjoy1127) [reply shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/advanced/TranslatableFederationExecutor.java:69]: No, we can't. They use different type of schema and table.
Finally we will keep only one translatable federation executor, and filterable related executor will be deleted.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java (hunk 11 above)
Comment (boyjoy1127) [reply shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/translatable/TranslatableTableScan.java:140]: The comment is useful.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/FederationExecutorFactory.java [hunk 14]
```diff
@@ -34,20 +34,20 @@ public final class FederationExecutorFactory {
     
     /**
      * Create new instance of federation executor factory.
-     * 
-     * @param databaseName database name
-     * @param schemaName schema name
+     *
+     * @param databaseName       database name
```
[the remark this answers is not in the document]
Comment (boyjoy1127) [reply shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-executor/src/main/java/org/apache/shardingsphere/infra/federation/executor/FederationExecutorFactory.java:38]: Yes, I use default IDEA setting.
Screen Shot

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**